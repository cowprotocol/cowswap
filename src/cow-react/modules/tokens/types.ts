import { CurrencyAmount, Token } from '@uniswap/sdk-core'

export interface OnchainState<T> {
  value: T
  loading: boolean
  syncing: boolean
  error: boolean
  valid: boolean
}

export type OnchainTokenAmount = OnchainState<CurrencyAmount<Token> | undefined>

export type TokenAmounts = {
  [tokenAddress: string]: OnchainTokenAmount
}

export type TokenAmountsResult = {
  amounts: TokenAmounts
  isLoading: boolean
}

export interface BalancesAndAllowancesParams {
  account: string | undefined
  spender: string | undefined
  tokens: Token[]
  blocksPerFetchBalance?: number
  blocksPerFetchAllowance?: number
}

export interface BalancesAndAllowances {
  balances: TokenAmounts
  allowances: TokenAmounts
  isLoading: boolean
}
