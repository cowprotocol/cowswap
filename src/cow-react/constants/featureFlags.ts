export const AMOUNTS_FORMATTING_FEATURE_FLAG = 'highlight-amounts-formatting'
export const limitOrdersFeatures = {
  // Temporary hide estimated execution price because it requires rework and retest
  DISPLAY_EST_EXECUTION_PRICE: false,
  DISPLAY_EXECUTION_TIME: false,
}
