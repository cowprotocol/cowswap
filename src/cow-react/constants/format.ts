export const DECIMAL_SEPARATOR = '.'
