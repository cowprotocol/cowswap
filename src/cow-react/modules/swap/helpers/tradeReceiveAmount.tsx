import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
import TradeGp from 'state/swap/TradeGp'
import { TokenAmount } from '@cow/common/pure/TokenAmount'
import { ReactNode } from 'react'

export interface ReceiveAmountInfo {
  type: 'from' | 'to'
  amountBeforeFees: ReactNode
  amountAfterFees: ReactNode
  amountAfterFeesRaw: CurrencyAmount<Currency>
  feeAmount: ReactNode
}

export function getInputReceiveAmountInfo(trade: TradeGp): ReceiveAmountInfo {
  return {
    type: 'from',
    amountBeforeFees: (
      <TokenAmount
        amount={
          trade.tradeType === TradeType.EXACT_INPUT && trade.inputAmountWithFee.lessThan(trade.fee.amount)
            ? null
            : trade.inputAmountWithoutFee
        }
        defaultValue="0"
      />
    ),
    amountAfterFeesRaw: trade.inputAmountWithFee,
    amountAfterFees: <TokenAmount amount={trade.inputAmountWithFee} defaultValue="0" />,
    feeAmount: <TokenAmount amount={trade.fee.feeAsCurrency} defaultValue="0" />,
  }
}

export function getOutputReceiveAmountInfo(trade: TradeGp): ReceiveAmountInfo {
  return {
    type: 'to',
    amountBeforeFees: <TokenAmount amount={trade.outputAmountWithoutFee} defaultValue="0" />,
    amountAfterFeesRaw: trade.outputAmount,
    amountAfterFees: <TokenAmount amount={trade.outputAmount} defaultValue="0" />,
    feeAmount: <TokenAmount amount={trade.outputAmountWithoutFee?.subtract(trade.outputAmount)} defaultValue="0" />,
  }
}
