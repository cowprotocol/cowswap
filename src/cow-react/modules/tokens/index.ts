export * from './types'

// Exported to allow to get the balance when the allowance is not needed (i.e. in token selection)
export { useOnchainBalances } from './hooks/useOnchainBalances'
export type { OnchainAllowancesParams } from './hooks/useOnchainBalances'

// Exported for all other cases, when we need the effective balance (i.e. )
export * from './hooks/useBalancesAndAllowances'
