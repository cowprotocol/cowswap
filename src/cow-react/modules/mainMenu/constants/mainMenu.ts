import { Routes } from '@cow/constants/routes'
import { CONTRACTS_CODE_LINK, DISCORD_LINK, DOCS_LINK, DUNE_DASHBOARD_LINK, TWITTER_LINK } from 'constants/index'
import { BasicMenuLink, InternalLink, MainMenuItemId, MenuItemKind, MenuTreeItem } from '../types'

// Assets
import IMAGE_DOCS from 'assets/cow-swap/doc.svg'
import IMAGE_TERMS_AND_CONDITIONS from 'assets/cow-swap/terms-and-conditions.svg'
import IMAGE_COOKIE_POLICY from 'assets/cow-swap/cookie-policy.svg'
import IMAGE_PRIVACY_POLICY from 'assets/cow-swap/privacy-policy.svg'
import IMAGE_INFO from 'assets/cow-swap/info.svg'
import IMAGE_CODE from 'assets/cow-swap/code.svg'
import IMAGE_DISCORD from 'assets/cow-swap/discord.svg'
import IMAGE_TWITTER from 'assets/cow-swap/twitter.svg'
import IMAGE_PIE from 'assets/cow-swap/pie.svg'
import IMAGE_SLICER from 'assets/cow-swap/ninja-cow.png'
import IMAGE_GAME from 'assets/cow-swap/game.gif'

export const isBasicMenuLink = (item: any): item is BasicMenuLink => {
  return !!(item.title && item.url)
}

export const FAQ_MENU: InternalLink[] = [
  { id: MainMenuItemId.FAQ_OVERVIEW, title: 'Overview', url: Routes.FAQ },
  { id: MainMenuItemId.FAQ_PROTOCOL, title: 'Protocol', url: Routes.FAQ_PROTOCOL },
  { id: MainMenuItemId.FAQ_TOKEN, title: 'Token', url: Routes.FAQ_TOKEN },
  { id: MainMenuItemId.FAQ_TRADING, title: 'Trading', url: Routes.FAQ_TRADING },
  { id: MainMenuItemId.FAQ_AFFILIATE, title: 'Affiliate', url: Routes.FAQ_AFFILIATE },
  { id: MainMenuItemId.FAQ_LIMIT_ORDERS, title: 'Limit orders', url: Routes.FAQ_LIMIT_ORDERS },
  { id: MainMenuItemId.FAQ_ETH_FLOW, title: 'Selling Native tokens', url: Routes.FAQ_ETH_FLOW },
]

export const ACCOUNT_MENU: InternalLink[] = [
  { id: MainMenuItemId.ACCOUNT_OVERVIEW, title: 'Overview', url: Routes.ACCOUNT },
  { id: MainMenuItemId.ACCOUNT_TOKENS, title: 'Tokens', url: Routes.ACCOUNT_TOKENS },
]

export const MAIN_MENU: MenuTreeItem[] = [
  {
    kind: MenuItemKind.DROP_DOWN,
    title: 'Trade',
    items: [
      {
        links: [
          { id: MainMenuItemId.SWAP, kind: MenuItemKind.DYNAMIC_LINK, title: 'Swap', url: Routes.SWAP },
          {
            id: MainMenuItemId.LIMIT_ORDERS,
            kind: MenuItemKind.DYNAMIC_LINK,
            title: 'Limit orders',
            url: Routes.LIMIT_ORDER,
          },
        ],
      },
    ],
  },
  {
    kind: MenuItemKind.DROP_DOWN,
    title: 'Account',
    items: [
      {
        links: ACCOUNT_MENU,
      },
    ],
  },
  {
    kind: MenuItemKind.DROP_DOWN,
    title: 'FAQ',
    items: [
      {
        links: FAQ_MENU,
      },
    ],
  },
  {
    kind: MenuItemKind.DROP_DOWN,
    title: 'More',
    items: [
      {
        sectionTitle: 'Overview',
        links: [
          {
            id: MainMenuItemId.MORE_DOCUMENTATION,
            title: 'Documentation',
            url: DOCS_LINK,
            iconSVG: IMAGE_DOCS,
            kind: MenuItemKind.EXTERNAL_LINK,
          },
          { id: MainMenuItemId.MORE_ABOUT, title: 'About', url: Routes.ABOUT, iconSVG: IMAGE_INFO },
          {
            id: MainMenuItemId.MORE_STATISTICS,
            title: 'Statistics',
            url: DUNE_DASHBOARD_LINK,
            iconSVG: IMAGE_PIE,
            kind: MenuItemKind.EXTERNAL_LINK,
          },
          {
            id: MainMenuItemId.MORE_CONTRACT,
            title: 'Contract',
            url: CONTRACTS_CODE_LINK,
            iconSVG: IMAGE_CODE,
            kind: MenuItemKind.EXTERNAL_LINK,
          },
        ],
      },
      {
        sectionTitle: 'Community',
        links: [
          {
            id: MainMenuItemId.MORE_DISCORD,
            title: 'Discord',
            url: DISCORD_LINK,
            iconSVG: IMAGE_DISCORD,
            kind: MenuItemKind.EXTERNAL_LINK,
          },
          {
            id: MainMenuItemId.MORE_TWITTER,
            title: 'Twitter',
            url: TWITTER_LINK,
            iconSVG: IMAGE_TWITTER,
            kind: MenuItemKind.EXTERNAL_LINK,
          },
        ],
      },
      {
        sectionTitle: 'Other',
        links: [
          { kind: MenuItemKind.DARK_MODE_BUTTON },
          { id: MainMenuItemId.OTHER_COW_RUNNER, title: 'CoW Runner', url: Routes.PLAY_COWRUNNER, icon: IMAGE_GAME },
          { id: MainMenuItemId.OTHER_MEV_SLICER, title: 'MEV Slicer', url: Routes.PLAY_MEVSLICER, icon: IMAGE_SLICER },
          {
            id: MainMenuItemId.OTHER_TERMS_AND_CONDITIONS,
            title: 'Terms and Conditions',
            url: Routes.TERMS_CONDITIONS,
            iconSVG: IMAGE_TERMS_AND_CONDITIONS,
          },
          {
            id: MainMenuItemId.OTHER_COOKIE_POLICY,
            title: 'Cookie Policy',
            url: Routes.COOKIE_POLICY,
            iconSVG: IMAGE_COOKIE_POLICY,
          },
          {
            id: MainMenuItemId.OTHER_PRIVACY_POLICY,
            title: 'Privacy Policy',
            url: Routes.PRIVACY_POLICY,
            iconSVG: IMAGE_PRIVACY_POLICY,
          },
        ],
      },
    ],
  },
]
