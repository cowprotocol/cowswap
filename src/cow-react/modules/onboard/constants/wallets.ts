import injectedModule from '@web3-onboard/injected-wallets'
import { SupportedChainId } from 'constants/chains'

export const WALLET_STORAGE_KEY = 'connectedWallets'

export const appMetadata = {
  name: 'CoW Swap',
  icon: '/images/cow-logo.svg',
  description: 'Moooooooooooo',
  recommendedInjectedWallets: [
    { name: 'MetaMask', url: 'https://metamask.io' },
    { name: 'Coinbase', url: 'https://wallet.coinbase.com/' },
  ],
}

const WC_BRIDGE_URL = 'https://bridge.walletconnect.org'

export const defaultWCOptions = {
  bridge: WC_BRIDGE_URL,
  qrcodeModalOptions: {
    mobileLinks: ['rainbow', 'metamask', 'argent', 'trust', 'imtoken', 'pillar'],
  },
  connectFirstChainId: true,
}

export const zenGoOptions = {
  ...defaultWCOptions,
  label: 'ZenGo',
  getIcon: async () => {
    console.log('debug called zengo')
    return (await import('../assets/icons/zengo.svg')).default
  },
}

export enum WalletOptions {
  INJECTED = 'injected',
  COINBASE = 'coinbase',
  WALLET_CONNECT = 'wallet_connect',
  ZENGO = 'zengo',
  KEYSTONE = 'keystone',
  LEDGER = 'ledger',
}

interface DisabledWallets {
  [key: string]: WalletOptions[]
}

export const DISABLED_WALLETS_BY_CHAIN: DisabledWallets = {
  [SupportedChainId.MAINNET]: [],
  [SupportedChainId.GNOSIS_CHAIN]: [WalletOptions.ZENGO],
  [SupportedChainId.GOERLI]: [WalletOptions.ZENGO, WalletOptions.COINBASE],
  // [SupportedChainId.POLYGON]: [],
}

export const CONNECT_OPTIONS = {
  [WalletOptions.INJECTED]: () => Promise.resolve(injectedModule()),
  [WalletOptions.COINBASE]: () => import('@web3-onboard/coinbase').then((m) => m.default()),
  [WalletOptions.WALLET_CONNECT]: () => import('@web3-onboard/walletconnect').then((m) => m.default(defaultWCOptions)),
  [WalletOptions.ZENGO]: () => import('@web3-onboard/walletconnect').then((m) => m.default(zenGoOptions)),
  [WalletOptions.KEYSTONE]: () => import('@web3-onboard/keystone').then((m) => m.default()),
  [WalletOptions.LEDGER]: () => import('@web3-onboard/ledger').then((m) => m.default()),
}
