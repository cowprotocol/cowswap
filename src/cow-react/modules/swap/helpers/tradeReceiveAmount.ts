import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
import { formatSmartAmount } from 'utils/format'
import TradeGp from 'state/swap/TradeGp'

export interface ReceiveAmountInfo {
  type: 'from' | 'to'
  amountBeforeFees: string
  amountAfterFees: string
  amountAfterFeesRaw: CurrencyAmount<Currency>
  feeAmount: string
}

export function getInputReceiveAmountInfo(trade: TradeGp): ReceiveAmountInfo {
  return {
    type: 'from',
    amountBeforeFees:
      trade.tradeType === TradeType.EXACT_INPUT && trade.inputAmountWithFee.lessThan(trade.fee.amount)
        ? '0'
        : formatSmartAmount(trade.inputAmountWithoutFee) || '0',
    amountAfterFeesRaw: trade.inputAmountWithFee,
    amountAfterFees: formatSmartAmount(trade.inputAmountWithFee) || '0',
    feeAmount: formatSmartAmount(trade.fee.feeAsCurrency) || '0',
  }
}

export function getOutputReceiveAmountInfo(trade: TradeGp): ReceiveAmountInfo {
  return {
    type: 'to',
    amountBeforeFees: formatSmartAmount(trade.outputAmountWithoutFee) || '0',
    amountAfterFeesRaw: trade.outputAmount,
    amountAfterFees: formatSmartAmount(trade.outputAmount) || '0',
    feeAmount: formatSmartAmount(trade.outputAmountWithoutFee?.subtract(trade.outputAmount)) || '0',
  }
}
