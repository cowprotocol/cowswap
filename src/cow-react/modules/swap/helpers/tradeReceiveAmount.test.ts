import { getInputReceiveAmountInfo, getOutputReceiveAmountInfo } from './tradeReceiveAmount'
import TradeGp from 'state/swap/TradeGp'
import { COW, GNO } from 'constants/tokens'
import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { CurrencyAmount, Price, Token, TradeType } from '@uniswap/sdk-core'
import { ReactElement } from 'react'

const currency = COW[SupportedChainId.MAINNET]
const currencyOut = GNO[SupportedChainId.MAINNET]
const amount = 250
const output = 250

function currencyAmount18e(currency: Token, amount: number): CurrencyAmount<Token> {
  return CurrencyAmount.fromRawAmount(currency, amount * 10 ** 18)
}

describe('Helpers to build ReceiveAmountInfo', () => {
  describe('getInputReceiveAmountInfo()', () => {
    describe('When selling, if the fee is bigger than the traded amount', () => {
      it('Then amountBeforeFees should be zero', () => {
        const value = getInputReceiveAmountInfo(
          new TradeGp({
            inputAmount: currencyAmount18e(currency, amount),
            inputAmountWithFee: currencyAmount18e(currency, 0),
            inputAmountWithoutFee: currencyAmount18e(currency, amount),
            outputAmount: currencyAmount18e(currency, output),
            outputAmountWithoutFee: currencyAmount18e(currency, output - 3),
            fee: { feeAsCurrency: currencyAmount18e(currency, 3), amount: '5000' },
            executionPrice: new Price(currency, currencyOut, 1, 4),
            tradeType: TradeType.EXACT_INPUT,
            quoteId: 10000,
          })
        )

        expect(value.amountAfterFeesRaw.toExact()).toBe('0')
        expect((value.amountAfterFees as ReactElement).props).toMatchSnapshot()
        expect((value.amountBeforeFees as ReactElement).props).toMatchSnapshot()
        expect((value.feeAmount as ReactElement).props).toMatchSnapshot()
        expect(value.type).toBe('from')
      })
    })
  })

  describe('getOutputReceiveAmountInfo()', () => {
    it('Should match a snapshot', () => {
      const value = getOutputReceiveAmountInfo(
        new TradeGp({
          inputAmount: currencyAmount18e(currency, amount),
          inputAmountWithFee: currencyAmount18e(currency, 0),
          inputAmountWithoutFee: currencyAmount18e(currency, amount),
          outputAmount: currencyAmount18e(currency, output),
          outputAmountWithoutFee: currencyAmount18e(currency, output - 3),
          fee: { feeAsCurrency: currencyAmount18e(currency, 3), amount: '5000' },
          executionPrice: new Price(currency, currencyOut, 1, 4),
          tradeType: TradeType.EXACT_INPUT,
          quoteId: 10000,
        })
      )

      expect(value.amountAfterFeesRaw.toExact()).toBe('250')
      expect((value.amountAfterFees as ReactElement).props).toMatchSnapshot()
      expect((value.amountBeforeFees as ReactElement).props).toMatchSnapshot()
      expect((value.feeAmount as ReactElement).props).toMatchSnapshot()
      expect(value.type).toBe('to')
    })
  })
})
