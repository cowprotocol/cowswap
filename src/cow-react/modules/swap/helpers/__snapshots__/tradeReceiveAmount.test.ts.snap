// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers to build ReceiveAmountInfo getInputReceiveAmountInfo() When selling, if the fee is bigger than the traded amount Then amountBeforeFees should be zero 1`] = `
Object {
  "amount": CurrencyAmount {
    "currency": Token {
      "address": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
      "chainId": 1,
      "decimals": 18,
      "isNative": false,
      "isToken": true,
      "name": "CoW Protocol Token",
      "symbol": "COW",
    },
    "decimalScale": JSBI [
      660865024,
      931322574,
    ],
    "denominator": JSBI [
      1,
    ],
    "numerator": JSBI [],
  },
  "defaultValue": "0",
}
`;

exports[`Helpers to build ReceiveAmountInfo getInputReceiveAmountInfo() When selling, if the fee is bigger than the traded amount Then amountBeforeFees should be zero 2`] = `
Object {
  "amount": null,
  "defaultValue": "0",
}
`;

exports[`Helpers to build ReceiveAmountInfo getInputReceiveAmountInfo() When selling, if the fee is bigger than the traded amount Then amountBeforeFees should be zero 3`] = `
Object {
  "amount": CurrencyAmount {
    "currency": Token {
      "address": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
      "chainId": 1,
      "decimals": 18,
      "isNative": false,
      "isToken": true,
      "name": "CoW Protocol Token",
      "symbol": "COW",
    },
    "decimalScale": JSBI [
      660865024,
      931322574,
    ],
    "denominator": JSBI [
      1,
    ],
    "numerator": JSBI [
      908853248,
      646484075,
      2,
    ],
  },
  "defaultValue": "0",
}
`;

exports[`Helpers to build ReceiveAmountInfo getOutputReceiveAmountInfo() Should match a snapshot 1`] = `
Object {
  "amount": CurrencyAmount {
    "currency": Token {
      "address": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
      "chainId": 1,
      "decimals": 18,
      "isNative": false,
      "isToken": true,
      "name": "CoW Protocol Token",
      "symbol": "COW",
    },
    "decimalScale": JSBI [
      660865024,
      931322574,
    ],
    "denominator": JSBI [
      1,
    ],
    "numerator": JSBI [
      933756928,
      902409669,
      216,
    ],
  },
  "defaultValue": "0",
}
`;

exports[`Helpers to build ReceiveAmountInfo getOutputReceiveAmountInfo() Should match a snapshot 2`] = `
Object {
  "amount": CurrencyAmount {
    "currency": Token {
      "address": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
      "chainId": 1,
      "decimals": 18,
      "isNative": false,
      "isToken": true,
      "name": "CoW Protocol Token",
      "symbol": "COW",
    },
    "decimalScale": JSBI [
      660865024,
      931322574,
    ],
    "denominator": JSBI [
      1,
    ],
    "numerator": JSBI [
      24903680,
      255925594,
      214,
    ],
  },
  "defaultValue": "0",
}
`;

exports[`Helpers to build ReceiveAmountInfo getOutputReceiveAmountInfo() Should match a snapshot 3`] = `
Object {
  "amount": CurrencyAmount {
    "currency": Token {
      "address": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
      "chainId": 1,
      "decimals": 18,
      "isNative": false,
      "isToken": true,
      "name": "CoW Protocol Token",
      "symbol": "COW",
    },
    "decimalScale": JSBI [
      660865024,
      931322574,
    ],
    "denominator": JSBI [
      1,
    ],
    "numerator": JSBI [
      908853248,
      646484075,
      2,
    ],
  },
  "defaultValue": "0",
}
`;
